lib/utils/log/file_writer.go

```diff
@@ -0,0 +1,182 @@
+/*
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package log
+
+import (
+	"context"
+	"io"
+	"io/fs"
+	"log/slog"
+	"os"
+	"path/filepath"
+	"sync"
+	"sync/atomic"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/gravitational/trace"
+)
+
+var (
+	// fileSharedWriter global file shared writer with the ability to reopen the log file
+	// in the same mode and with the same flags, either by function call or by filesystem
+	// notification subscription on the changes.
+	fileSharedWriter *FileSharedWriter
+
+	// initLock protects global init of the file shared writer.
+	initLock sync.Mutex
+)
+
+// FileSharedWriter is similar to SharedWriter except that it requires a `os.File` instead of a `io.Writer`.
+// This is to allow the File reopen required by logrotate and similar tools.
+type FileSharedWriter struct {
+	logFile  atomic.Pointer[os.File]
+	watcher  atomic.Pointer[fsnotify.Watcher]
```

Comment:
What's the point of the atomics if every interaction is serialized behind a lock?

Reply:
yes would be better to return lock only, since atomic not cover all cases when we in parallel call `Write`, `Reopen`, this atomic must cover only `runWatcherFunc` so will add lock there